sigmastate/src/test/scala/sigmastate/PrettyPrintErgoTreeSpecification.scala

```diff
@@ -0,0 +1,443 @@
+    PrettyPrintErgoTree.prettyPrint(compile(code)) shouldBe
+      "{ ($1: (Coll[Byte], Coll[Byte])) => xor($1._1, $1._2) }"
+  }
+
+  property("exponentiate group"){
+    val code = "{ (x: (GroupElement, BigInt)) => x._1.exp(x._2) }"
+    PrettyPrintErgoTree.prettyPrint(compile(code)) shouldBe
+      "{ ($1: (GroupElement, BigInt)) => $1._1.exp($1._2) }"
+  }
+
+  property("multiply group"){
+    val code = "{ (x: (GroupElement, GroupElement)) => x._1.multiply(x._2) }"
+    PrettyPrintErgoTree.prettyPrint(compile(code)) shouldBe
+      "{ ($1: (GroupElement, GroupElement)) => $1._1.multiply($1._2) }"
+  }
+
+  property("neq"){
+    val code = 
+      """{ (x: Context) =>
+        |  x.dataInputs(0).R4[Coll[Byte]].get != x.SELF.propositionBytes
+        |}""".stripMargin
+    PrettyPrintErgoTree.prettyPrint(compile(code)) shouldBe
+      """{ ($1: Context) =>
+         |  ($1.dataInputs(0.toInt).R4[Coll[Byte]].get) != (SELF.propositionBytes)
+         |}""".stripMargin
+  }
+
+  property("logical not"){
+    val code = "{ (x: Boolean) => !x }"
+    PrettyPrintErgoTree.prettyPrint(compile(code)) shouldBe
+      "{ ($1: Boolean) => !$1 }"
+  }
+
+  property("negation"){
+    val code = "{ (x: Byte) => -x }"
+    PrettyPrintErgoTree.prettyPrint(compile(code)) shouldBe
+      "{ ($1: Byte) => !$1 }"
+  }
+
+  property("pre header"){
+    val code = "{ (x: PreHeader) => x.height }"
+    PrettyPrintErgoTree.prettyPrint(compile(code)) shouldBe
+      "{ ($1: PreHeader) => $1.height }"
+  }
+
+  property("nested ergo script"){
+    val code = 
+      """OUTPUTS.zip(INPUTS).zip(OUTPUTS).zip(INPUTS)
+        | .map({ (t: (((Box, Box), Box), Box)) =>
+        | t._1._2.value + t._2.value
+        | }).fold(0L, { (a: Long, v: Long) => a + v }) == 10""".stripMargin
+    // TODO: $1 is shadowed, look closer to fold: possible problem with clash for Long/Box values, consider having unique global variable table?
+    PrettyPrintErgoTree.prettyPrint(compile(code)) shouldBe
+      """|(
+         |  OUTPUTS.zip(INPUTS).zip(OUTPUTS).zip(INPUTS).map(
+         |    { ($1: (((Box, Box), Box), Box)) => $1._1._2.value + $1._2.value }
+         |  ).fold(0.toLong, { ($1: (Long, Long)) => $1._1 + $1._2 })
+         |) == (10.toLong)""".stripMargin
+  }
+
+  property("various types in function definition"){
+    val code =
+      """{ (x: (String, (Any, (AvlTree, (Header, Unit))))) =>
+         |  false
+         |}""".stripMargin
+    PrettyPrintErgoTree.prettyPrint(compile(code)) shouldBe
+      "{ ($1: (String, (Any, (AvlTree, (Header, Unit))))) => false }"
+  }
+
+  // TODO: This doesn't look right
+  property("deserialize"){
+    import scorex.util.encode.Base58
+    import sigmastate.serialization.ValueSerializer
+    import sigmastate.Values.ByteArrayConstant
+
+    val str = Base58.encode(ValueSerializer.serialize(ByteArrayConstant(Array[Byte](2))))
+    val code = s"""deserialize[Coll[Byte]]("$str")(0) == 2"""
```

Comment:
how `deserialize` is transformed by compile? 
You can construct ErgoTree explicitly and don't use compile at all.